R/combine_checkboxes.R

```diff
@@ -40,87 +41,84 @@
 combine_checkboxes <- function(supertbl,
                                tbl,
                                cols,
-                               values_to,
+                               names_prefix = "",
+                               names_suffix = NULL,
+                               names_sep = "_",
```

Comment:
Am I understanding correctly that these parameters control the names of the new columns and not how variable names (ex. `race___0`) are parsed into names and values?

A couple comments:

1. If we go with this I think we need to make it clearer that we're assuming the checkbox fields are in `name___value` format and not giving the user any control over how that's parsed.
2. I don't think the defaults are right. In this example the output col is called `_race`. The default should probably just produce `race`:
```r
db_label |>
  combine_checkboxes(
    "demographics",
    starts_with("race")
  )
```
3. I see your "Maybe" in the PR comment and also think `names_glue` would be super valuable if we can do it 😊

Reply:
> Am I understanding correctly that these parameters control the names of the new columns and not how variable names (ex. race___0) are parsed into names and values?

They control the structure of the names, but the names themselves come from `.value` in `get_metadata_spec()`, i.e. the field name prior to the `___` checkbox changes.

> If we go with this I think we need to make it clearer that we're assuming the checkbox fields are in name___value format and not giving the user any control over how that's parsed.

How about we just add a `check_metadata_fields_exist()` in `get_metadata_spec()` similar to what I have in the parent function for `check_fields_exist()`? If checkbox names are changed and they don't appear in the metadata `field_name` column, we can throw an error and suggestion. We should expect users don't manipulate the metadata tibbles, but we need the connection between the metadata tibble and the data tibble to remain intact. I think this supports our "if you change things, you need to take some responsibility for them" mindset.

> I don't think the defaults are right. In this example the output col is called _race. The default should probably just produce race

I'm open to changing it, but when I was thinking through outputs I was worried about clashing with other possibly existing column names. See this example, if we're ok with that being the default in the event of a clash then I can rework this.

<details>
  <summary>Click me</summary>

```r
data <- tibble(
  id = 1,
  prefix = "prefix",
  x___1 = TRUE,
  x___2 = FALSE,
  y___1 = TRUE,
  y___2 = TRUE,
  z___1 = FALSE,
  x = "val"
)

metadata <- tibble(
  field_name = c("id", "prefix", "x___1", "x___2", "y___1", "y___2", "z___1", "x"),
  field_type = c("text", "text", rep("checkbox", 5), "text"),
  select_choices_or_calculations = c(NA, NA, rep("1, A | 2, B", 4), "3, C", NA),
)

suprtbl <- tibble(
  redcap_form_name = "tbl",
  redcap_data = list(data),
  redcap_metadata = list(metadata)
) |>
  as_supertbl()

combine_checkboxes(supertbl = suprtbl,
                   tbl = "tbl",
                   cols = c(starts_with("x__"), starts_with("y"), "z___1"),
                   names_sep = "") %>%
  pull(redcap_data)

New names:
• `x` -> `x...8`
• `x` -> `x...9`
[[1]]
# A tibble: 1 × 11
     id prefix x___1 x___2 y___1 y___2 z___1 x...8 x...9 y        z    
  <dbl> <chr>  <lgl> <lgl> <lgl> <lgl> <lgl> <chr> <fct> <fct>    <fct>
1     1 prefix TRUE  FALSE TRUE  TRUE  FALSE val   A     Multiple NA   
```

</details>

> I see your "Maybe" in the PR comment and also think names_glue would be super valuable if we can do it 😊

_ugh_... Fine. I knew it was coming but figured I'll get the rest of this ironed out first.

Reply:
> How about we just add a `check_metadata_fields_exist()` in `get_metadata_spec()` similar to what I have in the parent function for `check_fields_exist()`? If checkbox names are changed and they don't appear in the metadata `field_name` column, we can throw an error and suggestion. We should expect users don't manipulate the metadata tibbles, but we need the connection between the metadata tibble and the data tibble to remain intact. I think this supports our "if you change things, you need to take some responsibility for them" mindset.

Okay I'm fine with this. I would also add something to the documentation noting the pattern we're looking for.

> I'm open to changing it, but when I was thinking through outputs I was worried about clashing with other possibly existing column names. See this example, if we're ok with that being the default in the event of a clash then I can rework this.

I would resolve this with a warning or possibly error if the fields already exist. `pivot_longer()` for example errors and directs the user to the `names_repair` parameter to provide a repair strategy:

```r
tibble(x=1:3, y=4:6, value = 10) |>
 pivot_longer(x:y)

#>Error in `tidyr::pivot_longer()`:
#>! Names must be unique.
#>✖ These names are duplicated:
#>  * "value" at locations 1 and 3.
#>ℹ Use argument `names_repair` to specify repair strategy.
```

That may be too sophisticated for us but we may actually be able to recreate that behavior pretty easily with `vctrs::vec_as_names()` which is referenced in the docs for the `names_repair` parameter.

> _ugh_... Fine. I knew it was coming but figured I'll get the rest of this ironed out first.

Haha if it ends up being too tricky that's fine!

Reply:
Alright added support for `names_repair` and `names_glue`. 

`names_glue` I'm still iffy on, the use case in the `pivot_wider()` documentation is a bit more complicated than I believe we can support here, but try the current set up out and let me know what you think.

Reply:
Agree that `pivot_wider()` supports more but I think we're still providing a lot value with what we have.

Imagine a user has a `meals` instrument with some checkboxes like this:

|field_name|
|--|
|breakfast___apple|
|breakfast___orange|
|breakfast___spinach|
|lunch___apple|
|lunch___orange|
|lunch___spinach|
|dinner___apple|
|dinner___orange|
|dinner___spinach|

They could do:

```r
supertbl |>
  combine_checkboxes(
    "meals",
    matches("breakfast|lunch|dinner"),
    names_glue = "checkbox_{.value}_all"
  ) |>
  combine_checkboxes(
    "meals",
    matches("breakfast|lunch|dinner") & matches("apple|orange"),
    names_glue = "checkbox_{.value}_fruit"
  )
```

How slick is that?